provider/core/src/resource.rs

```diff
@@ -142,57 +142,46 @@ impl ResourceKey {
+            Body0,
+            Slash,
+            Body1,
+            AtSign,
+            Body2,
         }
         use State::*;
         i = start;
         let mut state = Start;
         loop {
-            state = match (
-                state,
-                if i < end {
-                    Some(path.as_bytes()[i])
-                } else {
-                    None
-                },
-            ) {
-                (Start, Some(b'a'..=b'z' | b'A'..=b'Z' | b'0'..=b'9' | b'_' | b'=')) => AfterChar,
-                (Start, _) => return Err(("[a-zA-Z0-9=_]", i)),
-
-                (AfterChar, Some(b'a'..=b'z' | b'A'..=b'Z' | b'0'..=b'9' | b'_' | b'=')) => {
-                    AfterChar
+            let content = if i < end {
+                Some(path.as_bytes()[i])
+            } else {
+                None
+            };
+            state = match (state, content) {
+                (Start | Body0, Some(b'a'..=b'z' | b'A'..=b'Z' | b'0'..=b'9' | b'_' | b'=')) => {
+                    Body0
                 }
-                (AfterChar, Some(b'/')) => AfterSlash,
-                (AfterChar, _) => return Err(("[a-zA-z0-9=_/]", i)),
-
-                (AfterSlash, Some(b'a'..=b'z' | b'A'..=b'Z' | b'0'..=b'9' | b'_' | b'=')) => {
-                    AfterCharAfterSlash
+                (Body0, Some(b'/')) => Slash,
+                (Slash | Body1, Some(b'a'..=b'z' | b'A'..=b'Z' | b'0'..=b'9' | b'_' | b'=')) => {
+                    Body1
                 }
-                (AfterSlash, _) => return Err(("[a-zA-Z0-9=_]", i)),
-
-                (
-                    AfterCharAfterSlash,
-                    Some(b'a'..=b'z' | b'A'..=b'Z' | b'0'..=b'9' | b'_' | b'='),
-                ) => AfterCharAfterSlash,
-                (AfterCharAfterSlash, Some(b'/')) => AfterSlash,
-                (AfterCharAfterSlash, Some(b'@')) => AfterAt,
-                (AfterCharAfterSlash, _) => return Err(("[a-zA-z0-9=_/@]", i)),
-
-                (AfterAt, Some(b'0'..=b'9')) => AfterDigit,
-                (AfterAt, _) => return Err(("[0-9]", i)),
-
-                (AfterDigit, Some(b'0'..=b'9')) => AfterDigit,
-                (AfterDigit, Some(_)) => return Err(("[0-9]", i)),
-                (AfterDigit, None) => {
+                (Body1, Some(b'/')) => Slash,
+                (Body1, Some(b'@')) => AtSign,
+                (AtSign | Body2, Some(b'0'..=b'9')) => Body2,
+
+                // Success:
+                (Body2, None) => {
                     return Ok(Self {
                         path,
                         hash: ResourceKeyHash::compute_from_str(path),
                     })
                 }
+
+                // Errors:
+                (Start | Slash, _) => return Err(("[a-zA-Z0-9=_]", i)),
+                (Body0, _) => return Err(("[a-zA-z0-9=_/]", i)),
+                (Body1, _) => return Err(("[a-zA-z0-9=_/@]", i)),
+                (AtSign | Body2, _) => return Err(("[0-9]", i)),
```

Comment:
This would make it even more structured and nicer imho:

```suggestion
            state = match state {
                Start => match byte {
                  Some(b'a'..=b'z' | b'A'..=b'Z' | b'0'..=b'9' | b'_' | b'=') => AfterChar,
                  _ => return Err(("[a-zA-Z0-9=_]", i)),
                },
                
                AfterChar => match byte {
                  Some(b'a'..=b'z' | b'A'..=b'Z' | b'0'..=b'9' | b'_' | b'=') => AfterChar
                  Some(b'/') => AfterSlash,
                  _ => return Err(("[a-zA-z0-9=_/]", i)),
                }

                AfterSlash => match byte {
                  Some(b'a'..=b'z' | b'A'..=b'Z' | b'0'..=b'9' | b'_' | b'=') => AfterCharAfterSlash
                  _ => return Err(("[a-zA-Z0-9=_]", i)),
                }

                AfterCharAfterSlash => match byte {
                  Some(b'a'..=b'z' | b'A'..=b'Z' | b'0'..=b'9' | b'_' | b'=') => AfterCharAfterSlash,
                  Some(b'/') => AfterSlash,
                  Some(b'@') => AfterAt,
                  _ => return Err(("[a-zA-z0-9=_/@]", i)),
                }

                AfterAt => match byte {
                  Some(b'0'..=b'9') => AfterDigit,
                  _ => return Err(("[0-9]", i)),
                }

                AfterDigit => match byte {
                  Some(b'0'..=b'9') => AfterDigit,
                  Some(_) => return Err(("[0-9]", i)),
                  None => {
                    return Ok(Self {
                        path,
                        hash: ResourceKeyHash::compute_from_str(path),
                    })
                }
              }
```
The error states are much more obviously correct because they're the fallthrough from the lines directly above.

Reply:
I find this more readable than what you did in #1540, but this still has the problem of the `b'a'..=b'z' | b'A'..=b'Z' | b'0'..=b'9' | b'_' | b'='` and stuff being duplicated.